tls/s2n_client_hello.c

```diff
@@ -69,8 +69,11 @@ static S2N_RESULT s2n_generate_client_session_id(struct s2n_connection *conn)
         return S2N_RESULT_OK;
     }
 
-    /* Only generate the session id for TLS1.3 if in middlebox compatibility mode */
-    if (conn->client_protocol_version >= S2N_TLS13 && !s2n_is_middlebox_compat_enabled(conn)) {
+    /* Only generate the session id for TLS1.3 if in middlebox compatibility mode
+     *
+     * actual_protocol_version is used here so the session id is generated if TLS1.3
+     * was not negotiated, even if the client supports it */
+    if (conn->actual_protocol_version >= S2N_TLS13 && !s2n_is_middlebox_compat_enabled(conn)) {
```

Comment:
Isn't this triggered [when generating the ClientHello](https://github.com/aws/s2n-tls/blob/32c2c0391065f1380176049466fd05fe5950f18c/tls/s2n_client_hello.c#L590)? At that point, we haven't negotiated a version with the server yet, so I think client_protocol_version and actual_protocol_version should be the same. actual_protocol_version is just set early to the same value as client_protocol_version for historic reasons.

Reply:
Ah, [this](https://github.com/aws/s2n-tls/blob/main/tls/s2n_resume.c#L219) might be the issue. It sets actual_protocol_version without setting client_protocol_version. I'd worry what other parts of our code assume actual_protocol_version==client_protocol_version before we actually negotiate a version :\

Reply:
Would it be better if I used `s2n_connection_get_protocol_version`? That returns `actual_protocol_version` as long as its not `S2N_UNKNOWN_PROTOCOL_VERSION`

Reply:
I think I'm just worried that there will be other bugs like this. But setting client_protocol_version too when you load a session isn't right either (and I think would disable downgrade protection).

So I guess we just need to fix the bug we know about.